eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ga4gh/GA4GHBeaconWSServerV2.java

```diff
@@ -0,0 +1,298 @@
+            }
+        }
+
+        return null;
+    }
+
+    public List<DatasetAlleleResponse> getDatasetAlleleResponsesHelper(List<VariantMongo> variantMongoList,
+                                                                       BeaconAlleleRequestBody request) {
+
+        List<DatasetAlleleResponse> datasetAllelResponses = new ArrayList<DatasetAlleleResponse>();
+
+        if (request.getIncludeDatasetResponses() == null ||
+                request.getIncludeDatasetResponses().equalsIgnoreCase("NONE")) {
+            return null;
+        }
+
+        List<VariantSourceMongo> variantSourceMongoList = service.findAllVariantSourcesForBeacon();
+
+        HashSet<String> studiesPresent = new HashSet<String>();
+        HashMap<String, Float> studyIdToFrequencyMapper = new HashMap<>();
+        variantMongoList.forEach(variantMongo -> {
+            variantMongo.getSourceEntries().forEach(variantSourceEntryMongo -> {
+                studiesPresent.add(variantSourceEntryMongo.getStudyId());
+            });
+            variantMongo.getVariantStatsMongo().forEach(variantStatisticsMongo -> {
+                if (variantMongo.getAlternate().equalsIgnoreCase(variantStatisticsMongo.getMafAllele())) {
+                    studyIdToFrequencyMapper.put(variantStatisticsMongo.getStudyId(), variantStatisticsMongo.getMaf());
+                }
+            });
+        });
+
+        HashMap<String, VariantSourceMongo> allStudies = new HashMap<>();
+        variantSourceMongoList.forEach(variantSourceMongo -> {
+            allStudies.put(variantSourceMongo.getStudyId(), variantSourceMongo);
+        });
+
+        allStudies.forEach((studyId, variantSourceMongo) -> {
+            if (studiesPresent.contains(studyId)) {
+                if (request.getIncludeDatasetResponses().equalsIgnoreCase("ALL") ||
+                        request.getIncludeDatasetResponses().equalsIgnoreCase("HIT")) {
+                    datasetAllelResponses.add(buildDatasetAlleleResponseHelper(true,
+                            variantSourceMongo,
+                            studyIdToFrequencyMapper.get(variantSourceMongo.getStudyId())==null?
+                                    null:new Float(studyIdToFrequencyMapper.get(variantSourceMongo.getStudyId()))));
+                }
+            } else {
+                if (request.getIncludeDatasetResponses().equalsIgnoreCase("ALL") ||
+                        request.getIncludeDatasetResponses().equalsIgnoreCase("MISS")) {
+                    datasetAllelResponses.add(buildDatasetAlleleResponseHelper(false,
+                            variantSourceMongo, null));
+                }
+            }
+        });
+
+        return datasetAllelResponses;
+    }
+
+    public  DatasetAlleleResponse buildDatasetAlleleResponseHelper(boolean exists, VariantSourceMongo variantSourceMongo, Float frequency) {
+        return new DatasetAlleleResponse(variantSourceMongo.getStudyId(),
+                exists,
```

Comment:
this and some other fit in the previous line

Reply:
fixed :+1: